File: Previous Rewrite/CombatTracker.DomainOld/Current/Actions/PsychicAction.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using CombatTracker.Domain.Reference;
using CombatTracker.Domain.Reference.Magic;

namespace CombatTracker.Domain.Current.Actions
{
    public class PsychicAction : BaseAction
    {
        public int CurrentRound { get; set; }
        public int TotalRounds { get; set; }
        public int LevelDifference { get; set; }


        public PsychicAction(RmssDbContext db, Actor theChar, double currentTime, int psychicLevel)
            : base(currentTime)
        {
            WhoIsActing = theChar;
            LevelDifference = theChar.Level - psychicLevel;
            TotalRounds = PsychicRefractoryPeriodChart.LookupRoundsToZero(db, LevelDifference);
            Name = "Psychic Recovery";
            Note = "Round " + CurrentRound + " of " + TotalRounds;
            BasePercent = 1;
            StartTime = currentTime;
            EndTime = theChar.CalculateTimeRequiredForSpells(MySettings.BaseRoundTime) + currentTime;
            theChar.CritNegatives += PsychicRefractoryPeriodChart.LookupNegative(db, LevelDifference, CurrentRound);
            ActionType = ActionTypeEnum.Effect;
            Type = ActorActionType.Psychic;
        }

        public override bool CharacterAction
        {
            get { return false; }
        }

        public override void HandleAction(RmssDbContext db, double currentTime)
        {
            WhoIsActing.CritNegatives -= PsychicRefractoryPeriodChart.LookupNegative(db, LevelDifference, CurrentRound);
            CurrentRound += 1;
            WhoIsActing.CritNegatives += PsychicRefractoryPeriodChart.LookupNegative(db, LevelDifference, CurrentRound);
            EndTime = WhoIsActing.CalculateTimeRequiredForSpells(MySettings.BaseRoundTime) + currentTime;
            Note = "Round " + CurrentRound + " of " + TotalRounds;
        }

        public override void RefreshData()
        {
            
        }
        public override bool Reoccuring
        {
            get { return CurrentRound <= TotalRounds; }
        }
    }
}

File: CombatTracker-Application/CombatTrackerData/Actions/PsychicAction.vb
Partial Class PsychicAction
    Public Sub New(ByVal DB As RMSSDataDataContext, ByVal TheChar As Actor, ByVal CurrentTime As Double, ByVal PsychicLevel As Integer)
        WhoIsActing = TheChar
        LevelDifference = TheChar.Level - PsychicLevel
        TotalRounds = PsychicRefractoryPeriodChart.LookupRoundsToZero(DB, LevelDifference)
        Me.Name = "Psychic Recovery"
        Me.Note = "Round " & CurrentRound & " of " & TotalRounds
        Me.BasePercent = 1
        Me.StartTime = CurrentTime
        Me.EndTime = TheChar.CalculateTimeRequiredForSpells(BaseRoundTime) + CurrentTime
        TheChar.CritNegatives += PsychicRefractoryPeriodChart.LookupNegative(DB, LevelDifference, CurrentRound)
        Me.ActionType = ActionTypeEnum.Effect
    End Sub
    Public Overrides ReadOnly Property CharacterAction() As Boolean
        Get
            Return False
        End Get
    End Property

    'Public Overrides Function Clone() As Action
    '    Dim Act As New PsychicAction
    '    Me.CloneMe(Act)
    '    Return Act
    'End Function

    Public Overrides Sub HandleAction(ByVal DB As RMSSDataDataContext, ByVal GI As Game)
        WhoIsActing.CritNegatives -= PsychicRefractoryPeriodChart.LookupNegative(DB, LevelDifference, CurrentRound)
        CurrentRound += 1
        WhoIsActing.CritNegatives += PsychicRefractoryPeriodChart.LookupNegative(DB, LevelDifference, CurrentRound)
        Me.EndTime = WhoIsActing.CalculateTimeRequiredForSpells(BaseRoundTime) + GI.CurrentTime
        Me.Note = "Round " & CurrentRound & " of " & TotalRounds

    End Sub
    Public Overrides ReadOnly Property Reoccuring() As Boolean
        Get
            Return CurrentRound <= TotalRounds
        End Get
    End Property
    Public Overrides Sub RefreshData()

    End Sub

    Private Sub OnCreated()

        Me.BasePercent = 0.001
        Name = ""
        StartTime = 0
        EndTime = 0
        Note = ""
        State = ActionProblem.IsOK
        CurrentModifier = 0
        Type = ActorActionType.Psychic
    End Sub
End Class
